Refactor test and coverage logic (#1542)
stryker-mutator/stryker-net: docs/technical-reference/Testing Framework.md

# Testing Frameworks undocumented
This document captures the discoveries made about the various testing frameworks while working on Stryker. 
This can help understanding the design and logic of the mutation test classes.
This is a mandatory reading for anyone trying to understand and/or fix bugs related to coverage capture, analysis and testing phase.

To be specific, this document will mostly (and implicitly) refer to their VsTest adapters.

## Generalities
The [VsTest protocol](https://github.com/Microsoft/vstest-docs/blob/master/RFCs/0006-DataCollection-Protocol.md) is not

> rich enought to cover evey test framework features. Therefore, each framework makes their own decision on how to handle

Reviewer: ```suggestion
rich enough to cover every test framework feature. Therefore, each framework makes their own decisions on how to handle
```

some of their specificities. That is the reason why we can observe different behaviour and surprising results.

> This affect:

Reviewer: ```suggestion
This affects:
```

- how tests are discovered
- how tests are described/identified
- how tests are filtered
- how tests are executed
- how test results are reported

## Identification

> By default, VsTest generates unique identifier stored as `Guid`. I do not know if test runners are able to provide

Reviewer: "unique identifier" -> "a unique identifier" or "unique identifiers"

a specific implementation but in practice, this identifier is derived from the(test's) display's name hash code.

## xUnit

### Theories
xUnit's theory is a pattern where some test is executed multiple time with varying input data (like NUnit's TestCase). 
These test mesthods accept one or more parameters and bear one of this attributes: `InlineData`, `MemberData` or `ClassData`.
There are differences
between compile time theories (data is fixed at build time) and run time theories (data is discovered at run time).

#### Static theories
Static discoveries are seen and processed as different test cases.

#### Run time theories
Run time theories are discovered as one test, disregarding the number of underlying data set.

> There is one test result per dataset, all associated with the unique test id.

Author: the phrase is ambiguous, and its trickier than that. I'll propose another version

Here is a summarized timeline of tests execution:
```

>    xUnit runner calls data theory's data source to feach each test value(s)

Reviewer: feach -> fetch

TestCase start event
   xUnit run test with first set of data

>    xUnit reports test reult

Reviewer: reult -> result/results

TestCase end event
   xUnit run test with second set of data
   xUnit reports test reult
   xUnit run test with third set of data
   ...
   xUnit reports last test result
TestCase start event
   ...
```

> The difficuly here is that a lot happens between 'testcase end' and 'testcase start' event. 

Reviewer: difficuly -> difficulty

At coverage capture it is a problem, because there is a risk of spilling coverage information to the next tests.

> And during execution phase, it is diffcult to predict when the test will really be over. 

Reviewer: diffcult -> difficult

> Also, if a mutation ends up changing a test case name - typically by changing the result of `ToTring()`, this

Author: I tried to clarify it a bit more.

testcase will only run when running **all tests** and can no longer be executed in isolation.

### NUnit

### TestCases
NUnit's TestCases is a pattern where some test is executed multiple time with varying input data (like xUnit's TestCase).

> These test mesthods accept one or more parameters and bear one of this attributes (on top of the `[Test]` attribue):

Reviewer: ```suggestion
These test mesthods accept one or more parameters and bear one of these attributes (on top of the `[Test]` attribute):
```

`ToString()` method for each parameter (e.g: `Test(1,2)` for parameter 1 and 2).
Then each test is run separately and a test result is provided for each test.
**Caveat**: if the `ToString()` method is not overriden, multiple test cases will have the same name 
(eg: `Test(MyClass`). Then, there will be no way to distinguish between tests as well as properly associate test
results.
When this happens, these appear as a single test resulting in multiple outcomes in Visual Studio.

Here is a summarized timeline of tests execution:
```
   NUnit runner calls the TestCaseSource and gets every test cases.
TestCase start event
   NUnit run test with first set of data
   NUnit reports test reult
TestCase end event
TestCase start event
   NUnit run test with second set of data
   NUnit reports test reult
TestCase end event
   ...
   NUnit reports lasy test reult
TestCase end event
TestCase start event
   ...
```

The problems for Stryker are:
1. If test cases have the same name, we cannot distinguish between test cases, so it must be handled as a single test. This *single test* reports multiple test cases so Stryker needs to wait either for one failure or for each testcase

> to succeed to establish the correct status. The good news is that discovery should provide the number of result to wait for.

Reviewer: ```suggestion
to succeed to establish the correct status. The good news is that discovery should provide the number of results to wait for.
```

> 2. Every mutations encountered by the `TestCaseSource` method will be reported as covered by the first test case,

Reviewer: "Every mutations" -> "every mutation"/"all mutations"

whereas it is likely these mutations would need to be adequately associated with their realtive test case.
3. TestCaseSource is called **before** Testcase start event which means Stryker may set the active mutation too late. 
The only way to handle this is to force any concerned mutants to run in dedicated test sessions.

> 

Author: I have not looked into MsTest, but
MsTest is not the most use framework and I assume there is a good integration between MsTest and VsTest. In any case, the remediation strategies I used to work around those issues should also work for MsTest based assemblies.
And, I am not sure this document is of any use, beyond technical curiosity. 
Completing means adding a new test project using MsTest and intense debugging to understand how it works.
I think It can wait for a rainy day :-).

TestCase start event
   ...
```
The difficulty here is that a lot happens between `testcase end` and `testcase start` events. 
At coverage capture it is a problem, because there is a risk of spilling coverage information to the next tests: Stryker
can only capture coverage information on `testcase end` events, and only in association with the current running test.
So, every mutants covered between `testcase end` and `testcase start` will be associated with the **next** test.

And during execution phase, it is impossible to predict when the test will really be over, so it is diffult to
establish is the test was succesful.  

> Also, if a mutation ends up changing a test case name - typically by changing the result of `ToTring()`, it will change the

Reviewer 2: ```suggestion
Also, if a mutation ends up changing a test case name - typically by changing the result of `ToString()`, it will change the
```

Reviewer 2: I've never thought about this. Is it an idea to skip mutating ToString methods to prevent this?

Author: 1) I enclosed detection logic that ensure any such offending mutants will be run against all tests
2) it can only happens for theories, which raises many concerns about testing in general

So, I think we don't need it, and its not worth it :-)
